File: 3rd Semester/Advanced Programming Methods(Done)/Laboratory/src (BUX)/Controller/Ctrl.java
package Controller;
import Exceptions.*;
import Model.*;
import Model.Statements.IStmt;
import Model.Utils.*;
import Repo.IRepo;
import View.GUI.PrgListController;

import java.io.IOException;
import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.stream.Collectors;

public class Ctrl {
    private IRepo repo;
    private ExecutorService executor;

    public void addProgram(PrgState st){
        repo.addPrg(st);
    }

    private Map<Integer, Integer> conservativeGarbageCollector(Collection<Integer> symTbl, Map<Integer, Integer> heap)
    {
        return heap.entrySet().stream().filter(e->symTbl.contains(e.getKey())).collect(Collectors.toMap(Map.Entry::getKey,
                Map.Entry::getValue));
    }

    public Ctrl(){}

    public Ctrl(IRepo r){repo=r;}

    private List<PrgState> removeCompletedPrg(List<PrgState> inPrgList) {
        return inPrgList.stream()
                .filter(p -> p.isNotCompleted())
                .collect(Collectors.toList());
    }


    public IRepo getRepo(){return this.repo;}


    private void oneStepForAllPrg(List<PrgState> prgList) throws InterruptedException {
        prgList.forEach(prg ->repo.logPrgStateExec(prg));


        List<Callable<PrgState>> callList = prgList.stream()
                .map((PrgState p) -> (Callable<PrgState>)(() -> p.oneStep()))
                .collect(Collectors.toList());


        List<PrgState> newPrgList = executor.invokeAll(callList). stream()
                . map(future -> {
                    try {
                        return future.get();
                    } catch (Exception e) {
                        System.out.println("\n!!!Empty Stack!!!");
                        return null;
                    }
                })
                .filter(p -> p!=null)
                .collect(Collectors.toList());
        prgList.addAll(newPrgList);

        prgList.forEach(prg ->repo.logPrgStateExec(prg));
        prgList.forEach(prg->System.out.println(prg.toString()));
        repo.setPrgList(prgList);

    }


    public void allStep() throws InterruptedException {
        executor = Executors.newFixedThreadPool(2);
        List<PrgState> prgList=removeCompletedPrg(repo.getPrgList());
        MyIList<Integer> out=new MyList<>();
        MyIHeap<Integer> heap=new MyHeap<>();
        while(prgList.size() > 0){

           Collection <Integer> symValues = new ArrayList<>();
            repo.getPrgList().forEach(e->symValues.addAll(e.getSymTable().getContent().values()));
            out=prgList.get(0).getOut();
            heap.setHeap(conservativeGarbageCollector(symValues,prgList.get(0).
                    getHeap().getContent()));
            oneStepForAllPrg(prgList);
            prgList = removeCompletedPrg(repo.getPrgList());
        }
        repo.CloseFile();
        executor.shutdownNow();
        repo.setPrgList(prgList);
        System.out.println(out.toString());
        System.out.println(heap.toString());
        System.out.println("End of program.");
    }

    public void allStepGUI() throws MyException {
        executor = Executors.newFixedThreadPool(2);
        List<PrgState> prgList = removeCompletedPrg(repo.getPrgList());
        if(prgList.size() == 0) {
            executor.shutdownNow();
            throw new MyException("Program finished.");
        }
        else {
            try {
                oneStepForAllPrg(prgList);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            executor.shutdownNow();
        }
    }
    public String toString() {
        return repo.getPrgList().get(0).getOriginalProgram().toString();
    }

}

File: 3rd Semester/Advanced Programming Methods/Laboratory/Toy Interpreter C#/sharplab/Controller/Ctrl.cs
using sharplab.Model.Statements;
using sharplab.Model.Utils;
using sharplab.Repo;
using System;
using System.Collections.Generic;
using System.Text;

namespace sharplab.Controller
{
    class Ctrl
    {
        private IRepo repo;

        public Ctrl(IRepo r) { repo = r; }


        private void OneStep(PrgState state)
        {
            IMyStack<IStmt> stk = state.GetStk();
            try
            {
                stk.IsEmpty();
            }
            catch
            {
                throw new Exception("Empty stack!\n");
            }
            IStmt crtStmt = stk.Pop();
                Console.WriteLine(crtStmt.ToString()+" ");
                crtStmt.Execute(state);

        }

        public void AllStep()
        {
            PrgState prg = repo.GetCrtPrg();
            try
            {
                while (true)
                {
                   OneStep(prg);
                   repo.LogPrgStateExec();
                   Print();
                }
            }
            catch
            {
                System.Console.WriteLine("End of program.");
            }
        }

        public void Print()
        {
            PrgState prog = repo.GetCrtPrg();
            IMyDictionary<string, int> dict = prog.GetSymTable();
            IMyList<int> list = prog.Geto();
            System.Console.WriteLine("-------------------------------------------\n");
            System.Console.WriteLine("SYM TABLE:\n"+dict.ToString() + "\n");
            System.Console.WriteLine("OUT:\n"+list.ToString() + "\n");
            System.Console.WriteLine("-------------------------------------------\n");
        }
    }
}
